llm/README.md

```diff
@@ -0,0 +1,71 @@
+# Personal CHTC Submit Template for LLM Fine-Tuning
+
+Use Case: Fine-tuning large language models on CHTC and track it with `Weights & Biases`.
+
+![WANDB](wandb.png)
+
+## Quick start
+
+1. Store your WANDB credentials and `STAGING_DIR` path in an environment file named .env. See the provided [example](.env.example).
+1. Update the `run_name` in `run.sh`. This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.
+1. Modify `run.sub` as necessary.
+1. SSH to the submit node.
+1. Create a `condor_log` directory using the command: `mkdir condor_log` if you don't have it.
+1. Submit your job using `condor_submit run.sub`.
+
+## Used Stacks
+
+- Docker
+- Github Container Registry (ghcr.io)
+- Huggingface Transformers
+- Weights & Biases (WANDB)
+
+## Used CHTC/HTCondor Features
+
+- Docker Universe
+- Checkpointing
+- Staging (for storing checkpoints)
+- GPU
+
+## Building your own container
+
+Note: Perform this step on your local machine, not on a CHTC submit node.
+
+Example resources for building a training container:
+
+- [Dockerfile](Dockerfile)
+- [requirements.txt](requirements.txt)
+- [Helper script](build.sh)
+- [.env](.env.example) for Github container registry credentials (`CR_PAT`)
+
+Users should consider building their own container to match their specific needs.
```

Comment:
If this step is optional, we can move this to the top and explicitly note it as an optional step.

Reply:
I should also mention that when I tested building my own container, a) it is a whopping 22GB and b) github defaulted to making it private so that my job couldn't fetch it. Is the token in `.env` supposed to be passed to condor somehow so it can fetch the ghcr container? Otherwise, we should mention the step of making sure the container is updated to public. I did verify that the token has `write/read: packages` set, so it should have the permission it needs to get the container.

@ChristinaLK As for the container being quite large, which is better -- using condor to move a large container over the network by pulling straight from ghcr.io, or should we set up `/staging/gpu-examples` like we did with `/squid/gpu-examples` for moving other large execution environments? (I assume 22GB is too large for squid to handle)

Reply:
> it is a whopping 22GB

How did it get so big? The latest version of `huggingface/transformers-pytorch-gpu` has a compressed size of 10.38 GB on DockerHub. I would expect it already had all the large dependencies like CUDA, PyTorch, etc.

https://hub.docker.com/r/huggingface/transformers-pytorch-gpu/tags shows recent versions that are smaller, only 8 GB.

Reply:
Yes you are right, the [Dockerfile](https://github.com/JasonLo/templates-GPUs/blob/b9797b8572e029dd1e2ba0400d8d72aec07e4b49/llm/Dockerfile) only install a few small python dependencies on top of the original `huggingface/transformers-pytorch-gpu`. I had tried to rebuild it today, but the size is still around 22 GB... Maybe they are using some more aggressive compression? Feel free to swap my image to any CHTC provided one if you see fit.